src/main/java/com/salesforce/dynamodbv2/mt/util/CachingAmazonDynamoDbStreams.java

```diff
@@ -783,6 +827,15 @@ public GetShardIteratorResult getShardIterator(GetShardIteratorRequest request)
         String dynamoDbIterator;
         switch (ShardIteratorType.fromValue(request.getShardIteratorType())) {
             case TRIM_HORIZON:
+                CachingShardIterator cachedIterator =
+                    trimHorizonCache.getIfPresent(new StreamShardId(request.getStreamArn(), request.getShardId()));
+                if (cachedIterator != null) {
+                    return new GetShardIteratorResult().withShardIterator(cachedIterator.toExternalString());
+                } else {
+                    dynamoDbIterator = loadShardIterator(request);
+                    getShardIteratorUncached.increment();
+                }
+                break;
```

Comment:
Instead of always looking up a physical iterator for trim horizon, it now tries to return a virtual one if the sequence number of the first record in the stream is know. There is a bit of a risk of handing this iterator back and then having trim horizon move forward. So if the iterator is used later and we no longer have those trimmed records in the cache, the client could get a trim horizon exception. I think KCL retries in that case though.

Reply:
Maybe add this as a comment in some form.

Reply:
Done https://github.com/salesforce/mt-dynamo/pull/467/commits/2e5b888851be86764fcc27db29d25a5ad81a254d